### PR TITLE
Implement a workaround to make build operation traces fully functional for composite builds.

### PR DESCRIPTION
**Pull request by ldaley**
Relates to #3873.

https://builds.gradle.org/project.html?projectId=Gradle&tab=projectOverview&branch_Gradle=ldaley%2Fcomposite-build-operation-trace-quickfix

### CODE REVIEW

**Comment by breskeby** [comment]
👍  lgtm.

**Comment by ldaley** [comment]
@breskeby please go ahead and merge it.

**Comment by ldaley** [comment]
Issue for the fundamental problem this works around: https://github.com/gradle/gradle/issues/4241

File: subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java [hunk 1]
```diff
@@ -112,9 +115,32 @@ public BuildOperationTrace(StartParameter startParameter, BuildOperationListener
...
+        // This is a workaround for https://github.com/gradle/gradle/issues/3873
+        // Several early typed operations have `buildPath` property,
+        // the value of which can only be determined after the settings file for the build has loaded.
+        //
+        // The workaround is to buffer all operation notifications in memory until the root build's settings have loaded.
+        // This works because all possible settings files have been evaluated by the time the root one has been.
+        // This is not guaranteed to hold into the future.
+        // A proper solution would be to change the operation details/results to be
+        // truly immutable and convey values known at the time.
+        listenerManager.addListener(new BuildAdapter() {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java]: ❌ It seems this listener is never removed, which could lead to a memory leak.

File: subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationRecord.java [hunk 2]
```diff
@@ -18,13 +18,32 @@
...
 import com.google.common.collect.Lists;
+import com.google.common.collect.Ordering;
 
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 
 public final class BuildOperationRecord {
 
+    public static final Ordering<BuildOperationRecord> ORDERING = Ordering.natural()
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationRecord.java]: What's the reason behind adding this ordering?

File: subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java (hunk 1 above)
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java]: I'll have a look on this

Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationTrace.java]: added a fix for this in https://github.com/gradle/gradle/pull/4278

File: subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationRecord.java (hunk 2 above)
Comment (ldaley) [reply subprojects/core/src/main/java/org/gradle/internal/operations/trace/BuildOperationRecord.java]: The sibling events are now ordered by who started first. Previously it was who finished first.

**Pull request merged**